### ISSUE TITLE
Error: Cannot connect to postgresql (problems with postmaster.pid and conflicting versions of postgresql)

### ISSUE THREAD

**Issue by elsstrange** [makersacademy/common-knowledge#258]
## TLDR
If you see an error like:

```
Is the server running locally and accepting connections on Unix domain socket "/tmp/.s.PGSQL.5432"?
```

Then try `tail /usr/local/var/log/postgres.log` to see your errors, and google them to get solutions.

I found the solutions to my particular errors in these resources:

- https://thoughtbot.com/blog/macos-postgres-could-not-connect-to-server
- https://stackoverflow.com/questions/17822974/postgres-fatal-database-files-are-incompatible-with-server

## THE WHOLE STORY

Working on the Acebook challenge, trying to execute the `bin/rails db:create` set-up command, I was getting this error:

```
could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.5432"?
Couldn't create database for {"adapter"=>"postgresql", "encoding"=>"unicode", "pool"=>5, "database"=>"pgapp_development"}
rails aborted!
PG::ConnectionBad: could not connect to server: No such file or directory
	Is the server running locally and accepting
	connections on Unix domain socket "/tmp/.s.PGSQL.5432"?
```

Our cohort had seen something similar back at the start of the Modern Web Development module, where the issue had been a missing file in usr/local/var/postgres but this didn't seem to be the case this time.

**Coming to the rescue, this fab little article:** https://thoughtbot.com/blog/macos-postgres-could-not-connect-to-server

Like the article described `pgrep -l postgres` did not produce any results. Postgres isn't running, right?

So I tried `brew services start postgresql` - but hit a snag:

```
Service `postgresql` already started, use `brew services restart postgresql` to restart.
```

Hmmm... So I tried `brew services restart postgresql` - which stopped and started the service fine, but `pgrep -l postgres` still didn't show anything.

So, even though I'm pretty sure that postgresql **is** running, I try the article's next step for when it isn't: `tail /usr/local/var/log/postgres.log`

And, this is the interesting result:

```
2020-12-07 21:28:37.947 GMT [1799] FATAL:  database files are incompatible with server
2020-12-07 21:28:37.947 GMT [1799] DETAIL:  The data directory was initialized by PostgreSQL version 12, which is not compatible with this version 13.1.
```

Hmmm, not the error the article suggested...

```
--- FLASHBACK #1 ---
One night ago, I'm installing the requirements for the Acebook, and I have to install Ruby 2.5.0.
Homebrew decides to have a massive party updating applications all over the place (I really should run homebrew updates more often...)

```

I can't specifically remember seeing it, but I **bet** it upgraded postgresql.

But wait... everyone else in the cohort had to do that Ruby install, and had that massive homebrew update (we **all** really should run homebrew updates more often) and they didn't have errors creating and migrating the db.

Oh well, a quick google for the error about mismatched versions of database server and database files yields this helpful Stack Overflow: https://stackoverflow.com/questions/17822974/postgres-fatal-database-files-are-incompatible-with-server

The top answer suggests I simply try `brew postgresql-upgrade-database`

Ooooh, it's working, it's working, it's working!

Oh no, wait, it isn't:

```
Checking cluster versions                                   ok
pg_resetwal: error: lock file "postmaster.pid" exists
pg_resetwal: Is a server running?  If not, delete the lock file and try again.
The target cluster lacks some required control information:
  checkpoint next XID
  latest checkpoint next OID
  latest checkpoint next MultiXactId
  latest checkpoint next MultiXactOffset
  first WAL segment after reset
  float8 argument passing method
  maximum alignment
  block size
  large relation segment size
  WAL block size
  WAL segment size
  maximum identifier length
  maximum number of indexed columns
  maximum TOAST chunk size
  dates/times are integers?
  data checksum version

Cannot continue without required control information, terminating
Failure, exiting
Error: Upgrading postgresql data from 12 to 13 failed!
==> Removing empty postgresql initdb database...
==> Moving postgresql data back from /usr/local/var/postgres.old to /usr/loca
Error: Failure while executing; `/usr/local/opt/postgresql/bin/pg_upgrade -r -b /usr/local/Cellar/postgresql@12/12.5/bin -B /usr/local/opt/postgresql/bin -d /usr/local/var/postgres.old -D /usr/local/var/postgres -j 4` exited with 1.
```

YUCK.  But wait, what's that at the top of the error message?

```
pg_resetwal: error: lock file "postmaster.pid" exists
```

Didn't that first article mention this error? 

> FATAL:  lock file "postmaster.pid" already exists
> This indicates that Postgres had not been shut down cleanly, for any number of reasons. The first line of the postmaster.pid file is the process identifier of what once was Postgres.

Not shut down cleanly, eh?

```
--- FLASHBACK #2 ---
A week ago I'm just chilling with my Macbook. 

I close the screen, think of something I meant to do before finishing, reopen it and the screen goes bright white and unhappy.  

It's not responding to anything so I do a hard reboot. It starts up again and everything is fine.
```
OK, so it sounds like the postmaster.pid is also part of the problem.  I go in search.

Unfortunately the grep provided in the article doesn't find anything.  Nor does navigating to the usr/local/var/postgres directory.  I try the `rm` suggested by the article, and the terminal tells me the file doesn't exist.

This is where things get sketchy.  I try stopping and starting postgres a couple of times using homebrew to see if I can see this nasty little pid file using `ls`. No luck. By this point, in case it matters, my working directory is usr/local/var/postgres/

I decide to try `brew postgresql-upgrade-database` again to see if I can spot anything helpful in the error message.

Unexpectedly this time the upgrade works.  Somewhere along the way, I seem to have shaken off the postmaster.pid.  Sorry not to be more specific, but there really is nothing in my terminal logs that suggests when or how - maybe I should have just tried running the upgrade command a second time straight away.

Anyhow, `pgrep -l postgres` now works, and, more importantly, so does `bin/rails db:create`

# HURRAH!

Big thanks to the authors of that article and Stack Overflow answer. Thanks also to cohort buddies for earlier debugging attempts. Hope you enjoyed my documentation...
